### DISCUSSION TITLE
man page menu completion on another command

### DISCUSSION THREAD

**Discussion by mozirilla213** [akinomyoga/ble.sh#287] [Q&A] [answered]
I have a command `A` with a manpage `A` and ble.sh parses correctly the options when `A --<tab>` and gives a menu completion. I make an alias `alias B=A`, how can I make it so the command `B` shows the options of manpage `A` when `B --<tab>`?

Edit: I realized that could be workaround by magic expanding the alias with `bleopt edit_magic_expand=sabbrev:alias`, but my situation was actually: a command and manpage `A.sh` that has a wrapper sourceable function `B`. I cannot alias `B=A.sh` since B needs to run interactively

**Comment by akinomyoga** [comment]
Have you set `shopt -s progcomp_alias`? If not, please set it and see if it works.

**Reply by mozirilla213** [reply]
That solves the original situation, great.

But my issue actually was that the command and manpage have that extension, so the completions show on the executable `A.sh --<tab>`, but what I actually use is a function name differently `B --<tab>`, not the executable directly, an `alias B=A.sh` would not work interactively

**Reply by akinomyoga** [reply]
There are actually separate completion generations in `ble.sh`. One is `ble.sh`'s built-in one and the other is the integration with the programmable completion framework designed for Bash/Readline. The option `shopt -s progcomp_alias` only affects the behavior of the completion generation by Bash's programmable completion. If you see that `alias B=A.sh` doesn't work while `alias B=A` works, I don't think the behavior difference is caused by the extension. I guess there is a programmable completion setting for `A` but not for `A.sh`, which would make a difference.

I'll consider implementing `alias progcomp_alias` for `ble.sh`'s built-in completion generation.

**Reply by akinomyoga** [reply]
@mozirilla213 I now tried to implement it, but I realized that it is already implemented and works as expecetd.

Could you describe the setup, in which you are faced with the problem, in a bit more detail? In my environment, for example, the following setup works as expected:

```bash
$ _test1() { COMPREPLY=($(compgen -W 'hello world' -- "${COMP_WORDS[COMP_CWORD]}")); }
$ complete -F _test1 test1.sh
$ alias A=test1.sh
$ A [TAB]
```

**Reply by mozirilla213** [reply]
I'm sorry for my confusing wording, the situation goes something like this

```bash
$ export MANPATH=~/.local/man
$ mkdir -p ~/.local/man/man8

$ # The manpage was named "myscript.sh"
$ cat <<\EOF > ~/.local/man/man8/myscript.sh.8
.TH MYSCRIPT.SH 8 "myscript.sh Manual"
.SH SYNOPSIS
.SY runme
.SH DESCRIPTION
.PP
Change directory to /etc
.SH OPTIONS
.TP
.B \-h ", " \-\-help
Option -h
.TP
.B \-a ", " \-\-all
Option -a
EOF

$ PATH+=:~/.local/bin
$ mkdir -p ~/.local/bin
$ # The executable script was named "myscript.sh"
$ cat <<\EOF >  ~/.local/bin/myscript.sh
#!/usr/bin/env bash
echo "/etc"
EOF
$ chmod +x ~/.local/bin/myscript.sh

$ # It connects to a function that runs in current environment
$ # because it `cd`s, changes env variables, etc.
$ runme() {
  cd "$(myscript.sh)"
  export SCRIPT_PATH=$PWD
}

$ myscript.sh -<TAB>  # Menu completions from the man page appear
$ runme -<TAB>        # Menu completions don't appear of course
$ alias runme=myscript.sh
# Making an alias will generate completions but
# the functionality can't be replicated since the
# executable subprocess can't modify current interactive
# environment
```
I want to get the completions from the manpage in the function. I know a `_minimal` function gets createad when completing from the script, but I don't know how to redirect it to the manpage completions.

Edit: I updated the title to reflect the question

**Reply by akinomyoga** [reply]
Thank you for explaining the details about what is to be achieved! I'll reply in a separate comment.

**Comment by akinomyoga** [comment, marked as answer]
In commit 54ace59c57557169ce68366f23849cfe5f8413e8, I have extracted the related code as an independent function `ble/complete/source:option/generate-for-command`.

With this new interface, you can essentially do something like the following:

```bash
complete -F _comp_runme runme
_comp_runme() { ble/complete/source:option/generate-for-command myscript.sh "${comp_words[@]:1:comp_cword-1}"; }
```

More careful implementation of `_comp_runme` can be found [here](
https://github.com/akinomyoga/ble.sh/blob/467fa448a4766f5ee486af8ff74e852c40f011d4/memo/D2014-complete-mandb-for-command.bash#L63-L76).

The usage of `ble/complete/source:option/generate-for-command` is briefly explained in [the code comment](https://github.com/akinomyoga/ble.sh/commit/54ace59c57557169ce68366f23849cfe5f8413e8#diff-c180d07f9f6314cd0ce60494fd95e4a1a3afaa2f78afd7046897c4c4e6f072d3R5122-R5141).

**Reply by akinomyoga** [reply]
@mozirilla213 Also, could you test this?

**Reply by mozirilla213** [reply]
Yes, this is perfect, works on both examples, thank you for the work!

**Reply by akinomyoga** [reply]
Thank you.
